### DISCUSSION TITLE
[Typegoose] How to handle relations with MongoDB SubDocuments?

### DISCUSSION THREAD

**Discussion by guiaramos** [doug-martin/nestjs-query#1474] [Q&A]
Hi there,

sub documents are also know as embedded documents, this is the explanation on typegoose github:
```ts
class Job {
  @prop()
  public title?: string;

  @prop()
  public position?: string;
}

class Car {
  @prop()
  public model?: string;
}

class User {
  @prop()
  public name?: string;

  @prop({ required: true })
  public age!: number; // This is a single Primitive

  @prop({ type: () => [String] })
  public preferences?: string[]; // This is a Primitive Array

  @prop()
  public mainJob?: Job; // This is a single SubDocument

  @prop({ type: () => Job })
  public jobs?: Job[]; // This is a SubDocument Array

  @prop({ ref: () => Car })
  public mainCar?: Ref<Car>; // This is a single Reference

  @prop({ ref: () => Car })
  public cars?: Ref<Car>[]; // This is a Reference Array
}
```

I know that the library works well with **reference** of other documents:
```ts
  @Prop({ ref: () => MembershipEntity })
  memberships?: Ref<MembershipEntity>[]; // tyegoose Reference Array
```

But **I didn't find any explanation** on how to handle **subdocuments**:
```ts
  @Prop({ type: () => MembershipEntity })
  memberships?: MembershipEntity[]; // tyegoose SubDocument Array
```

I have tried to use the @CursorConnection:
```ts
@CursorConnection('memberships', () => MembershipDTO)
```

however, it does not work as expected because I **there is no resolver to add a new membership** subdoc to the array of memberships..

Also, when using normal type referencing:
```ts
  @Field(() => [MembershipDTO], { nullable: true })
  memberships?: MembershipDTO[];
```
I get the following error:
```bash
Error starting server, Error: Cannot determine a GraphQL input type for the "memberships". Make sure your class is decorated with an appropriate decorator.
```

**How to correct handle the subdocuments using Typegoose?**

**Comment by smolinari** [comment]
@guiaramos - There isn't a need to "handle" subdocuments with Mongoose. They are automatically part of the parent parent via dot notation, like `user.memberships`. What is it you are trying to accomplish that gives you that error? Can you show the resolver and entity/DTO code?

Scott

**Reply by guiaramos** [reply]
@smolinari first, thanks for your answer.

I am trying to accomplish the following:

This is my entities:
```ts
import { modelOptions, Prop } from '@typegoose/typegoose';

export class AddressEntity extends BaseIdEntity {
  @Prop({ required: true })
  nickname?: string;
  @Prop({ required: true })
  recipient?: string;
  @Prop({ required: true })
  phoneNumber?: string;
  @Prop({ required: true })
  address?: string;
  @Prop({ required: true })
  addressDetail?: string;
  @Prop({ required: true, default: false })
  default?: boolean;
}


@modelOptions({
  schemaOptions: {
    collection: 'users',
    timestamps: true,
  },
})
export class UserEntity extends BaseEntity {
  // ...
  @Prop({ type: () => AddressEntity })
  address?: AddressEntity[];
  // ...
}
```

This is my DTOS:
```ts
@ObjectType('Address')
export class AddressDTO extends BaseIdDTO {
  @FilterableField({ nullable: true })
  nickname?: string;
  @FilterableField({ nullable: true })
  recipient?: string;
  @FilterableField({ nullable: true })
  phoneNumber?: string;
  @FilterableField({ nullable: true })
  address?: string;
  @FilterableField({ nullable: true })
  addressDetail?: string;
  @FilterableField({ nullable: true })
  default?: boolean;
}

@ObjectType('User')
@UnPagedRelation('address', () => AddressDTO)
export class UserDTO extends BaseDTO {
  //...
}

```

The resolvers are the automatic ones "CRUDResolver".

on database I expect to have an array of address inside of user entity:
![Screen Shot 2021-11-10 at 5 11 34 PM](https://user-images.githubusercontent.com/49703106/141074945-0cec943d-bc84-4022-8ae7-da9999071f77.png)

without nestjs-query, I could make a a mutation which the input is the address fields and then use $push operator that will add a new address entity and generate the _id:
```ts
  async addAddressToUser(user: UserDTO, input: AddressEntity): Promise<User> {
      this.userModel
        .findOneAndUpdate({ cognitoId }, { $push: { address: input } }, { new: true })
        .exec()
  }
```

with nestjs-query,  I trying to add: `@UnPagedRelation('address', () => AddressDTO)`  but the input generated ask for `relationIds: [ID!]!` and not the address fields.


Could you please explain how can I accomplish this with nestjs-query? In other words, how can I $push an address entity to user entity, or which way you recommend me?

Thanks!

**Comment by smolinari** [comment]
Ok. So, `AddressDTO` needs to have `@InputType` and not `@ObjectType`. `@ObjectType` is only for output/ response objects. Basically any DTO needs `@InputType` as the decorator. 

Scott

**Reply by guiaramos** [reply]
hmm I got it, then nestjs-query creates a automatcally the `@InputType` for each primative/enums field of`UserDTO`  like string, number and etc.. but it doesn't for another `@ObjectType` correct? 

I need to make a custom ``@InputType`  for update the `UserDTO` with address field. is my understand correct?

**Reply by guiaramos** [reply]
I followed this path, but didnt succeed:

DTOs
```ts
@InputType('AddressUpdateInput')
export class AddressUpdateDTO {
  @Field({ nullable: true })
  nickname?: string;
  @Field({ nullable: true })
  recipient?: string;
  @Field({ nullable: true })
  phoneNumber?: string;
  @Field({ nullable: true })
  address?: string;
  @Field({ nullable: true })
  addressDetail?: string;
  @Field({ nullable: true })
  default?: boolean;
}

@InputType('UserUpdateInput')
export class UserUpdateDTO {
  @Field(() => [AddressUpdateDTO], { nullable: true })
  address?: AddressUpdateDTO[];
}

@InputType()
export class UpdateOneUserInput extends UpdateOneInputType(UserDTO, UserUpdateDTO) {}
```

Resolver
```ts
@Resolver(() => UserDTO)
export class UserResolver extends CRUDResolver(UserDTO, {
  delete: { disabled: true },
  UpdateDTOClass: UserUpdateDTO,
}) {

  constructor(readonly service: UserService, readonly niceService: NiceService) {
    super(service);
  }
]
```
did the input:
```json
{
  "input": {
    "id": "618baaab75db720506223b2d",
    "update": {
      "address": [
        {
          "nickname": "home",
          "recipient": "Gui",
          "phoneNumber": "00000000",
          "address": "home",
          "addressDetail": "home",
          "default": true
        }
      ]
    }
  }
}
```

and got the following error:
```bash
Unable to find reference addresses on UserEntity
```

**Reply by smolinari** [reply]
> hmm I got it, then nestjs-query creates a automatcally the `@InputType` for each primative/enums field of`UserDTO` like string, number and etc.. but it doesn't for another `@ObjectType` correct?
> 
> I need to make a custom ``@InputType` for update the`UserDTO` with address field. is my understand correct?

This has nothing to do with `nestjs-query`. You need to mark input types as such, so `nestjs-graphql` can create the correct types in the DSL. 

As for the new error, why is GraphQL looking for `addresses` and not `address`?

Scott

**Reply by guiaramos** [reply]
@smolinari yeah, i don't know why it's looking for addresses as there is no field with this name.. i will keep trying to find the way to work with this and will post here if success.. thanks for your help!
